### PR TITLE
[ESP32] Added support of door-lock for esp32

### LINKED ISSUE #24015: [BUG] [ESP32] Lock app cannot control Lock/Unlock
### Reproduction steps

```bash
https://github.com/project-chip/connectedhomeip/blob/master/examples/lock-app/esp32/main/DeviceCallbacks.cpp

this example support only on/off(`app::Clusters::OnOff::Id`).
I guess it must support `DoorLock::Id' and `lock/unlock`.

Google Nest hub2 cannot control Lock/Unlock of ESP32.
```

### Bug prevalence

always

### GitHub hash of the SDK that was being used

c76ba8f

### Platform

esp32

### Platform Version(s)

_No response_

### Anything else?

for example, PSoC6 example has a lot of callback function.
[https://github.com/project-chip/connectedhomeip/blob/master/examples/lock-app/infineon/psoc6/src/ZclCallbacks.cpp](https://github.com/project-chip/connectedhomeip/blob/master/examples/lock-app/infineon/psoc6/src/ZclCallbacks.cpp)

ESP32 lock app sample has only on/off function.

### PR DESCRIPTION
**Pull request by jadhavrohit924**
Problem

> ESP32 don’t have handling of door lock in lock-app.
> all cluster app using linux door-lock
> Fixes #24015.

Changes

> Added support of door lock for esp32

Testing

> Tested commissioning and door lock/unlock for esp32.

### CODE REVIEW

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp [hunk 1]
```diff
@@ -72,3 +76,19 @@ void AppDeviceCallbacks::OnOnOffPostAttributeChangeCallback(EndpointId endpointI
 exit:
     return;
 }
+
+void AppDeviceCallbacks::OnDoorLockPostAttributeChangeCallback(EndpointId endpointId, AttributeId attributeId, uint8_t * value)
+{
+    VerifyOrExit(attributeId == ZCL_LOCK_STATE_ATTRIBUTE_ID, ESP_LOGI(TAG, "Unhandled Attribute ID: '0x%04x", attributeId));
```
Comment (bzbarsky-apple) [inline examples/lock-app/esp32/main/DeviceCallbacks.cpp:82]: Please use the namespaced attribute ids, not the to-be-removed ZCL_* constants.

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp [hunk 2]
```diff
@@ -72,3 +76,19 @@ void AppDeviceCallbacks::OnOnOffPostAttributeChangeCallback(EndpointId endpointI
...
+    VerifyOrExit(endpointId == 1 || endpointId == 2, ESP_LOGE(TAG, "Unexpected EndPoint ID: `0x%02x'", endpointId));
+    if (*value)
+    {
+        BoltLockMgr().InitiateAction(AppEvent::kEventType_Lock, BoltLockManager::LOCK_ACTION);
```
Comment (bzbarsky-apple) [inline examples/lock-app/esp32/main/DeviceCallbacks.cpp:86]: Shouldn't this somehow pass through the endpoint id?  Which thing are we locking/unlocking?

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by wqx6** [comment]
I remember that STL is used in Linux door-lock code and STL is not acceptable in embedded code. Could you find a better way to add the door-lock support for ESP32?

Also ESP32 all-clusters-app uses the Linux door-lock code [here](https://github.com/project-chip/connectedhomeip/blob/master/examples/all-clusters-app/esp32/main/CMakeLists.txt#L33).

**Review by dhrishi — CHANGES_REQUESTED** [review]
@jadhavrohit924  Requesting changes to validate the comment from @wqx6

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 1 above)
Comment (jadhavrohit924) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: Done

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 2 above)
Comment (jadhavrohit924) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: We are passing endpoint in the callback `OnDoorLockPostAttributeChangeCallback ` itself

File: examples/lock-app/esp32/main/CMakeLists.txt [hunk 3]
```diff
@@ -147,6 +147,7 @@ idf_component_register(PRIV_INCLUDE_DIRS
                       "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/zzz_generated/lock-app/"
                       "${CMAKE_SOURCE_DIR}/third_party/connectedhomeip/examples/providers"                      
                       "${CMAKE_CURRENT_LIST_DIR}/include"
+		      "${IDF_PATH}/components/freertos/include/freertos"
```
Comment (wqx6) [inline examples/lock-app/esp32/main/CMakeLists.txt:150]: Why add this path to the PRIV_INCLUDE_DIRS? Can we add the freertos to the PRIV_REQUIRES components of the main component?

**Review by wqx6 — COMMENTED** [review]
Can the ESP32 all-clusters-app example use the lock manager here instead of the lock manager in Linux lock example?

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: Yes, I know, but we have two locks here: one on endpoint 1 and one on endpoint 2.  Why are we not passing which lock the action is for to the `InitiateAction` call?

File: examples/lock-app/esp32/main/CMakeLists.txt (hunk 3 above)
Comment (jadhavrohit924) [reply examples/lock-app/esp32/main/CMakeLists.txt]: Updated

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 2 above)
Comment (jadhavrohit924) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: This was redundency. Lock/Unlock action pasees through endpoints [here](https://github.com/project-chip/connectedhomeip/pull/24169/files#diff-eaa1b70782d02d5990bbfad4e85ae09441b53e9647185806e9c12840d0cfc1b0R49)

**Comment by jadhavrohit924** [comment]
@bzbarsky-apple @wqx6 @dhrishi Please review.

File: examples/lock-app/esp32/main/DeviceCallbacks.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: Maybe I misunderstand what `InitiateAction` does?  Why does it not need to know which endpoint's bolt is supposed to move?

Comment (jadhavrohit924) [reply examples/lock-app/esp32/main/DeviceCallbacks.cpp]: `InitiateAction` make sure that if one action is completed then only register other action.

**Review by dhrishi — APPROVED** [review]

File: examples/platform/esp32/lock/BoltLockManager.h [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    void SetAutoLockDuration(uint32_t aDurationInSecs);
+    bool NextState();
+    bool IsActionInProgress();
+    bool InitiateAction(int32_t aActor, Action_t aAction);
+
+    typedef void (*Callback_fn_initiated)(Action_t, int32_t aActor);
+    typedef void (*Callback_fn_completed)(Action_t);
+    void SetCallbacks(Callback_fn_initiated aActionInitiated_CB, Callback_fn_completed aActionCompleted_CB);
+
+    bool Lock(chip::EndpointId endpointId, const Optional<chip::ByteSpan> & pin, DlOperationError & err);
```
Comment (mlepage-google) [inline examples/platform/esp32/lock/BoltLockManager.h:147]: How does this build, given that it was renamed earlier in #24099 ?

**Comment by bzbarsky-apple** [comment]
It's a merge conflict.  The last time CI ran on this PR was before #24358 merged (3 days ago vs yesterday)...

We need to fix this new code ASAP to work with the right names; right now this is going to block all PRs from landing.

File: examples/platform/esp32/lock/BoltLockManager.h (hunk 4 above)
Comment (bzbarsky-apple) [reply examples/platform/esp32/lock/BoltLockManager.h:147]: It's a merge conflict.  The last time CI ran on this PR was before #24358 merged (3 days ago vs yesterday)...

We need to fix this new code ASAP to work with the right names; right now this is going to block all PRs from landing.

Comment (mlepage-google) [reply examples/platform/esp32/lock/BoltLockManager.h:147]: It's blocking me from merging master into my feature branch. Let me see about reverting this one.

Comment (mlepage-google) [reply examples/platform/esp32/lock/BoltLockManager.h:147]: Revert: https://github.com/project-chip/connectedhomeip/pull/24531

**Pull request merged**